doc: services: implement grouping of services docs

kartben opened the pull request:
https://builds.zephyrproject.io/zephyr/pr/100126/docs/services/index.html

Currently we have 39 (and not every "service" even has proper documentation at the moment...) top-level entries listed at the top level of [OS Services](https://docs.zephyrproject.org/latest/services/index.html) page. This list is virtually useless for anyone who does not already know whether the "service" or class of service they need even exists.

This PR overhauls the (OS) Services landing page by grouping these services into logical categories/buckets using a grid layout. This improves navigation and provides a better high-level overview of what Zephyr has to offer.

<img width="1214" height="1127" alt="image" src="https://github.com/user-attachments/assets/16f90c45-1506-4b47-80c9-b80073bb11e7" />

I have settled on 12 categories at this point, which one could say is still a lot, but it's important that the grouping doesn't become too vague/meaningless. 

Things that this PR does not necessarily address / A note on terminology:

- "Services" is a very broad term, and things that are (currently, as well as after this PR) listed under Services are a mixed bag of "libraries"/"utilities"/"subsystems"/"...". We certainly need to be better at terminology (see e.g. https://github.com/zephyrproject-rtos/zephyr/pull/100118) but in the meantime I don't think it is fundamentally wrong to call these things "services".

ljd42 commented:
Super Idea @kartben 

I have created something similar (albeit not that detailed) for presentation purposes. 

Find below some possible tinkering, in the case we want to reduce the number of categories to 8 or 9. Just throwing some ideas, it's not a change request :wink: 

**Device Management**
Everything we need to manage the device (boot/reboot, authentication, update)
- Cryptography
- Device Management
- Linkable Loadable Extensions (LLEXT)
- Task Watchdog
- "Security & Identity" could be included here too

**IPC & Messaging**
add Asynchronous notification?

**I/O and Data Processing**
Everything related to I/O processing
In addition to list proposed, move these services here:
- Real Time I/O (RTIO)
- Sensing Subsystem
- DSP

**System and Utilities**
- Binary Descriptors
- Resource Management
- Memory Attributes
- Network Buffers
- State Machine Framework (could also be listed in Device Management) 

**Remove**
- Algorithms 
- Application Services

fabiobaltieri reviewed (dismissed):
looking good

kartben commented:
Thanks for the feedback @mathieuchopstm and @henrikbrixandersen.

> > I must admit, my eyes _really_ prefer just a bulleted list like what we currently have here: [docs.zephyrproject.org/latest/services/index.html](https://docs.zephyrproject.org/latest/services/index.html)
> > While the squares, symbols, bold text, greyed text etc. may appear more "cool", it certainly doesn't make it easier to navigate (same goes for the recent redesign of the front, [docs.zephyrproject.org/latest](https://docs.zephyrproject.org/latest/)).
> 
> +1 on this: as proposed, if I know what I'm looking for (say, `LLEXT`) but not in which category it falls, I have to half-guess, click on a big square, _load a new page_ and hope my guess is correct.

Replying to the "[...] if I know what I'm looking for [...]" bit specifically: I would _really_ hope that people use built-in search (pro-tip: hit `/` key and start searching straight away, no mouse involved :)) as I think this will always be the fastest way (by far!) to navigate the docs for someone who know what they're looking for (kind of like how I hope one doesn't try to manually navigate the doxygen docs hierarchy to get to the docs of that function that they already know how it's called!).

I will definitely look into fixing the missing "+"s, as I had noticed that too. Having the buttons _and_ the bullet list is definitely a way to fix this but this comes with other Sphinx issues.

kartben committed:
doc: add sphinx-design extension

Enables grids, cards, and Material icons for documentation pages.

Assisted-by: Claude:opus-4.8
Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: add card styling for categorized index pages

Reuses the landing page's card design language (tinted card, Material
watermark icon, chevron links) with a masonry layout. Promotes the shared
shadow tokens to the global stylesheet.

Assisted-by: Claude:opus-4.8
Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: services: group OS Services into categorized cards

Replaces the flat list with category cards. Each card renders its category
page's toctree, keeping the service list in a single place (shown both on
the card and in the sidebar).

Assisted-by: Claude:opus-4.8
Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: services: move modem docs under connectivity

Adds a redirect and updates the MAINTAINERS entry.

Assisted-by: Claude:opus-4.8
Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

mathieuchopstm reviewed (commented):
Would it be possible to sort all entries within a group in alphabetical order?

e.g.
```
Connectivity
Networking
Bluetooth
USB
Controller Area Network (CAN) Bus Protocols
LoRa and LoRaWAN
Modbus
Modem modules
```

-->

```
Bluetooth
Connectivity
Controller Area Network (CAN) Bus Protocols
LoRa and LoRaWAN
Modbus
Modem modules
Networking
USB
```

Otherwise, this looks pretty nice.

kartben commented:
> Would it be possible to sort all entries within a group in alphabetical order?
> 
> [...]

@mathieuchopstm Yeah it's kind of an intentional "editorial" choice on this one, which I think might be worth keeping. Similarly I don't think we would necessarily want the "Logging, Tracing & Debugging" section to start with "CPU Load" ... but I appreciate this is a double-edged sword and may be hard to maintain and arbitrate going forward.

mathieuchopstm commented:
> > Would it be possible to sort all entries within a group in alphabetical order?
> > [...]
> 
> Yeah it's kind of an intentional "editorial" choice on this one, but I am happy with making this alphabetically sorted

For some sections, it *might* make sense to have a different order (e.g., `Device Management` being a the top of `Device Management`), but I can't really see one for `Connectivity` for example.

> (mind you it's impossible to enforce in CI with `zephyr-keep-sorted` tags though, because the file name of the rst document, used for the sorting, may be different from the actual title of the document)

True, but I think it would still be beneficial overall (e.g., `State Machine Framework` is `smf` - can go high a few ranks due to `'m' < 't'`, but still remains in the overall `S` group)

henrikbrixandersen commented:
Love the categories, looks good. I would prefer the UI not to "move" on mouse-over, though.

kartben commented:
> Love the categories, looks good.

Thanks!

> I would prefer the UI not to "move" on mouse-over, though.

Right, I think I _could_ be convinced to drop this UI choice but on the other hand that's the exact same styling we've had on the home page for a while now, so it's consistent :)

henrikbrixandersen commented:
> Right, I think I _could_ be convinced to drop this UI choice but on the other hand that's the exact same styling we've had on the home page for a while now, so it's consistent :)

Ugh, right. You can tell how seldom I actually use the links on the page as opposed to the sidebar links.

henrikbrixandersen commented on doc/services/frameworks.rst:
```diff
@@ -0,0 +1,14 @@
+
+   llext/index.rst
+   sensing/index.rst
+   smf/index.rst
```
SMF is not only for application-level code, perhaps it ought to go somewhere else - not sure where, though.

kartben replied:
Yeah LLEXT and SMF I was really tempted to put them under "Cool stuff, check it out" category. I'll think about it, and maybe other will chime in with ideas

henrikbrixandersen replied:
They are more like "frameworks"?

ljd42 commented:
Look great @kartben, Thanks!!
